rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+
+# RFC 005: Soft Chain Upgrades
+
+## Changelog
+
+- 2020-11-16: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+The current method for upgrading a network across block protocol breaking
+changes is to hard fork the chain. All nodes must coordinate a halt height/
+time, where application state can be exported across and an entirely new chain
+(as far as Tendermint is concerned) is created (see [cosmoshub-3
+upgrade](https://blog.cosmos.network/cosmos-hub-3-upgrade-announcement-39c9da941aee)).
+This means that each block protocol version maps to the entire life of a single
+chain. The result of this is that either Tendermint is hindered with the
+flexibility and quantity of changes it can make or that networks have to deal
+with pain of frequently upgrading.
+
+Furthermore, although work from [RFC002: Non-Zero Genesis](https://github.com/tendermint/spec/blob/master/rfc/002-nonzero-genesis.md)
+has enabled unique heights throughout the history of a "logical" chain, data
+retrievability across heights remains difficult as blockchains that choose to
+upgrade not only have to coordinate upgrades but they essentially lose all their
+history.
+
+The scope of this RFC is to address the need for Tendermint to have flexibility
+in improving itself whilst offering greater ease of use to the networks running
+on Tendermint. This is done by introducing a division in the types of ways a
+network can be upgraded. There will still remain the need for hard upgrades (the
+starting of a new chain) but this will not be focused on in this
+RFC(discussions on improving this can be found [here](https://github.com/tendermint/tendermint/issues/5595)).
+Instead, this RFC will only explore what supporting soft upgrades would entail.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+Soft and hard upgrades would directly align with semver. For example, a release
+from 1.2 → 1.3 would be considered a soft upgrade and could be executed on a
+live chain (more on how that would be executed further down). A major release
+from 1.5 → 2.0 would be a hard upgrade and require a network to halt and migrate.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block version consists of all block related data structures i.e. Header,
+Vote, Validator, Commit. This would be the sole focus for soft upgrades (a full
+list of the exact data structures can be found in
+[data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+A block version could be incremented between minor releases. However, for
+backwards compatibility, all prior versions in a given major release would also
+need to be supported thus TM 1.3 could support block version 9 and 10 and TM 1.4
+could introduce block version 11 but would still need to support 9 and 10. In a
+major release the support of prior versions could be dropped. Continuing with
```

Comment:
Could prior versions be kept around? I find this useful if we were to offer RPC querying across versions.

Reply:
Prior major versions you mean?

Reply:
prior block versions. If a chain started on block version 9 (TM 1.3) and is now on block version 12 (TM 2.3) as a user I may want to query a block from version 9.

Reply:
Across major versions i.e. 1.x -> 2.x would still be a hard fork and thus require a new chain (as is the case currently). I would also like to find ways to make hard forks easier on networks but I want to address this in a separate discussion

Reply:
> Across major versions i.e. 1.x -> 2.x would still be a hard fork and thus require a new chain (as is the case currently).

For p2p upgrades only right? It seems to me this upgrade process would mitigate hard forks for everything minus p2p upgrades.